File: polyglot/ruby/config_parser.rb
require 'yaml'
require 'json'
require 'date'
require 'uri'

module Dastlite
  # Base exception for all config-related errors
  class ConfigError < StandardError; end
  
  # Specific error types with helpful messages
  class MissingRequiredField < ConfigError
    def initialize(field:, context: "config")
      super("Missing required field '#{field}' in #{context}")
    end
  end

  class InvalidFormat < ConfigError
    def initialize(field:, expected:, actual:)
      msg = "Invalid format for '#{field}': expected #{expected}, got #{actual.inspect}"
      super(msg)
    end
  end

  class SecretNotResolved < ConfigError
    def initialize(secret_name:)
      super("Secret '#{secret_name}' not found in environment or vault")
    end
  end

  # Represents a single target URL/domain to scan
  Target = Struct.new(:url, :method, :headers) do
    def self.from_config(config)
      url = config['url'] || raise(InvalidFormat, new: 'url', expected: String, actual: config['url'])
      
      method = (config['method'] || :get).to_sym
      
      headers = if h = config['headers'] && !h.empty?
        h.transform_keys { |k| k.to_s.delete_prefix('X-').delete_prefix('x-') }
          .transform_values { |v| v.is_a?(String) ? v : YAML.safe_load(v, symbolize_names: true) }
      else
        {}
      end
      
      new(url, method, headers)
    end
    
    def to_safeguarded_url
      uri = URI.parse(self.url)
      "#{uri.scheme}://#{uri.host}#{uri.path}"
    end
  end

  # Authentication configuration
  AuthConfig = Struct.new(:type, :credentials) do
    def self.from_config(config)
      type = (config['type'] || 'basic').to_sym
      
      credentials = if c = config['credentials'] && !c.empty?
        case type
        when :basic
          { user: c['user'], pass: c['pass'] }
        when :bearer, :oauth2
          { token: c['token'] || c['access_token'] }
        when :api_key
          { key: c['key'], header: (c['header'] || 'X-API-Key') }
        else
          raise InvalidFormat.new(field: 'auth.type', expected: [:basic, :bearer, :oauth2, :api_key].join(', '), actual: type)
        end
      else
        {}
      end
      
      new(type, credentials)
    end
    
    def to_safeguarded_hash
      hash = { type: self.type.to_s }
      
      case self.type
      when :basic
        hash[:user] = self.credentials[:user] || ''
        hash[:pass] = self.credentials[:pass] || ''
      when :bearer, :oauth2
        hash[:token] = self.credentials[:token] || ''
      when :api_key
        hash[:key] = self.credentials[:key] || ''
        hash[:header] = self.credentials[:header] || 'X-API-Key'
      end
      
      hash
    end
  end

  # A single DAST rule from a ruleset
  Rule = Struct.new(:id, :name, :enabled, :params) do
    def self.from_config(config)
      id = config['id'] || raise(InvalidFormat.new(field: 'rule.id', expected: String, actual: config['id']))
      
      enabled = (config['enabled'] != false)
      
      params = if p = config['params'] && !p.empty?
        YAML.safe_load(p, symbolize_names: true) || {}
      else
        {}
      end
      
      new(id, config['name'], enabled, params)
    end
    
    def to_safeguarded_hash
      hash = { id: self.id, name: self.name, enabled: self.enabled }
      
      if !self.params.empty?
        hash[:params] = self.params.transform_keys(&:to_s).transform_values do |v|
          v.is_a?(String) ? YAML.dump(v) : v
        end
      end
      
      hash
    end
  end

  # A complete ruleset configuration
  RulesetConfig = Struct.new(:name, :version, :rules) do
    def self.from_config(config)
      name = config['name'] || 'default'
      version = (config['version'] || '1.0').to_s
      
      rules = if r = config['rules'] && !r.empty?
        r.map { |c| Rule.from_config(c) }
      else
        []
      end
      
      new(name, version, rules)
    end
    
    def to_safeguarded_hash
      hash = { name: self.name, version: self.version, rules: self.rules.map(&:to_safeguarded_hash) }
      hash
    end
  end

  # Main configuration parser and validator
  class Parser
    DEFAULT_CONFIG = {
      targets: [],
      auth: { type: :bearer },
      rulesets: [
        { name: 'default', version: '1.0', rules: [] }
      ],
      output: { format: :sarif, deduplicate: true },
      environment: { variables: {}, secrets: {} }
    }.freeze

    # Initialize with optional default config path
    def initialize(default_config_path = nil)
      @default_config = DEFAULT_CONFIG.dup
      
      if default_config_path && File.exist?(default_config_path)
        begin
          loaded = YAML.safe_load(File.read(default_config_path), symbolize_names: true)
          merge_configs(@default_config, loaded)
        rescue StandardError => e
          raise ConfigError.new("Failed to load default config from #{default_config_path}: #{e.message}")
        end
      end
    end

    # Parse and validate the main configuration file
    def parse(config_path)
      unless File.exist?(config_path)
        raise ConfigError.new("Config file not found: #{config_path}")
      end
      
      begin
        loaded = YAML.safe_load(File.read(config_path), symbolize_names: true) || DEFAULT_CONFIG.dup
        
        # Merge with defaults (user config overrides defaults)
        merge_configs(@default_config, loaded)
        
        # Validate and resolve secrets
        validate_and_resolve_secrets!
        
        @parsed = ConfigObject.new(merged)
      rescue StandardError => e
        raise ConfigError.new("Failed to parse config: #{e.message}")
      end
      
      @parsed
    end

    private

    def merge_configs(base, override)
      base.each do |key, value|
        if override.key?(key) && !override[key].nil?
          if value.is_a?(Hash) && override[key].is_a?(Hash)
            merge_configs(value, override[key])
          else
            base[key] = override[key]
          end
        end
      end
      
      # Apply overrides in reverse order (last wins for scalars)
      override.each do |key, value|
        if !base.key?(key) || (base[key].nil? && !value.nil?)
          base[key] = value
        elsif value.is_a?(Hash) && key == 'rulesets'
          # Merge rulesets arrays
          base[key] = merge_arrays(base[key], override[key])
        end
      end
      
      base
    end

    def merge_arrays(defaults, overrides)
      defaults.map do |d|
        o = overrides.find { |r| r['name'] == d['name'] } || {}
        ConfigObject.new(merge_configs(d, o))
      end
    end

    def validate_and_resolve_secrets!
      # Resolve environment variables in config values
      resolve_env_vars!(@parsed)
      
      # Validate required fields
      validate_targets!
      validate_rulesets!
    end

    def resolve_env_vars!(obj, prefix = '')
      obj.each do |key, value|
        next unless value.is_a?(String) || (value.is_a?(Hash) && !value.empty?)
        
        # Support ${VAR} and #{ENV['VAR']} syntax
        resolved_value = if value.match?(/\$\{([^}]+)\}/) || value.match?(/#\{\w+\}/)
          value.gsub(/\$|\{|\}/) { |m| ENV[m] || m }
        else
          value
        end
        
        # Recurse into nested hashes
        if resolved_value.is_a?(Hash) && !resolved_value.empty?
          resolve_env_vars!(ConfigObject.new(resolved_value), "#{prefix}.#{key}")
        elsif resolved_value.is_a?(Array)
          resolved_value.map! { |v| v.is_a?(String) ? v.gsub(/\$|\{|\}/) { |m| ENV[m] || m } : v }
        end
        
        obj[key] = resolved_value if resolved_value != value
      end
    rescue StandardError => e
      raise ConfigError.new("Failed to resolve environment variables: #{e.message}")
    end

    def validate_targets!
      targets = @parsed.targets || []
      
      unless targets.is_a?(Array)
        raise InvalidFormat.new(field: 'targets', expected: Array, actual: targets.class)
      end
      
      targets.each_with_index do |target, i|
        t = Target.from_config(target)
        
        unless URI.parse(t.url).is_a?(URI::HTTP) || URI.parse(t.url).is_a?(URI::HTTPS)
          raise InvalidFormat.new(field: "targets[#{i}].url", expected: 'http/https URL', actual: t.url)
        end
        
        @parsed.targets[i] = t if targets.is_a?(Array)
      end
      
      @parsed.targets || []
    end

    def validate_rulesets!
      rulesets = @parsed.rulesets || []
      
      unless rulesets.is_a?(Array)
        raise InvalidFormat.new(field: 'rulesets', expected: Array, actual: rulesets.class)
      end
      
      rulesets.each_with_index do |rs, i|
        r = RulesetConfig.from_config(rs)
        
        @parsed.rulesets[i] = r if rulesets.is_a?(Array)
      end
      
      @parsed.rulesets || []
    end

    # Wrapper class for the parsed configuration
    ConfigObject = Struct.new(:targets, :auth, :rulesets, :output, :environment) do
      def to_safeguarded_hash
        hash = {
          targets: self.targets.map(&:to_safeguarded_url),
          auth: self.auth.to_safeguarded_hash,
          rulesets: self.rulesets.map(&:to_safeguarded_hash),
          output: { format: self.output[:format], deduplicate: self.output[:deduplicate] }
        }
        
        if !self.environment.empty?
          hash[:environment] = {
            variables: self.environment[:variables],
            secrets: self.environment[:secrets]
          }
        end
        
        hash
      end

      def to_sarif_metadata
        # Return metadata for SARIF output
        {
          tool: { name: 'dastlite', version: '1.0' },
          driver_version: "config/#{self.rulesets.map(&:version).join(', ')}",
          targets: self.targets.map(&:to_safeguarded_url),
          auth_type: self.auth[:type].to_s,
          ruleset_count: self.rulesets.size
        }
      end

      def to_json(only_metadata: false)
        data = only_metadata ? to_sarif_metadata : to_safeguarded_hash
        JSON.generate(data)
      end
    end
  end
end

File: polyglot/python/config_parser.py
"""
polyglot/python/config_parser.py

A headless, config-as-code DAST runner that crawls an authenticated 
web/mobile-API surface and fires a curated active-scan ruleset, 
emitting deduplicated SARIF.

This module provides the configuration parser for dastlite.
"""

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Union

# Try to import pydantic for validation; fall back gracefully if missing
try:
    from pydantic import BaseModel, Field, validator, ValidationError
    PYDANTIC_AVAILABLE = True
except ImportError:
    PYDANTIC_AVAILABLE = False


@dataclass
class AuthConfig:
    """Authentication configuration."""
    type: str = "basic"  # basic, bearer, oauth2, ntlm
    username: Optional[str] = None
    password: Optional[str] = None
    token: Optional[str] = None
    scope: Optional[str] = None
    
    def to_dict(self) -> Dict[str, Any]:
        result = {}
        if self.type:
            result["type"] = self.type
        if self.username:
            result["username"] = self.username
        if self.password:
            result["password"] = self.password
        if self.token:
            result["token"] = self.token
        if self.scope:
            result["scope"] = self.scope
        return result


@dataclass
class TargetConfig:
    """Target URL configuration."""
    base_url: str
    paths: List[str] = field(default_factory=lambda: ["/"])
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "base_url": self.base_url,
            "paths": self.paths,
        }


@dataclass
class RulesetConfig:
    """Active-scan ruleset configuration."""
    enabled: bool = True
    categories: List[str] = field(default_factory=lambda: [
        "xss", 
        "csrf", 
        "sql-injection", 
        "broken-authentication"
    ])
    severity_threshold: str = "medium"  # low, medium, high
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "enabled": self.enabled,
            "categories": self.categories,
            "severity_threshold": self.severity_threshold,
        }


@dataclass
class OutputConfig:
    """Output format configuration."""
    formats: List[str] = field(default_factory=lambda: ["sarif"])
    output_dir: str = "."
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "formats": self.formats,
            "output_dir": self.output_dir,
        }


@dataclass
class RateLimitConfig:
    """Rate limiting configuration."""
    requests_per_second: float = 10.0
    burst_size: int = 50
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "requests_per_second": self.requests_per_second,
            "burst_size": self.burst_size,
        }


@dataclass
class Config:
    """Complete DASTLite configuration."""
    targets: TargetConfig = field(default_factory=TargetConfig)
    auth: AuthConfig = field(default_factory=AuthConfig)
    ruleset: RulesetConfig = field(default_factory=RulesetConfig)
    output: OutputConfig = field(default_factory=OutputConfig)
    rate_limit: RateLimitConfig = field(default_factory=RateLimitConfig)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "targets": self.targets.to_dict(),
            "auth": self.auth.to_dict(),
            "ruleset": self.ruleset.to_dict(),
            "output": self.output.to_dict(),
            "rate_limit": self.rate_limit.to_dict(),
        }


def _load_yaml_file(path: Path) -> Dict[str, Any]:
    """Load and parse a YAML file."""
    try:
        import yaml
    except ImportError:
        # Fallback to json if pyyaml not available
        with open(path) as f:
            return json.load(f)
    
    with open(path) as f:
        data = yaml.safe_load(f)
        return data if isinstance(data, dict) else {}


def _load_json_file(path: Path) -> Dict[str, Any]:
    """Load and parse a JSON file."""
    with open(path) as f:
        return json.load(f)


def _expand_env_vars(value: str) -> str:
    """Expand environment variables in a string value."""
    if not isinstance(value, str):
        return value
    
    result = os.path.expandvars(value)
    
    # Handle ${VAR} style for better compatibility
    import re
    def replace_var(match):
        var_name = match.group(1)
        env_val = os.environ.get(var_name, "")
        if not env_val:
            # Try to load from config file fallback
            return match.group(0)
        return env_val
    
    pattern = r'\$\{([^}]+)\}'
    result = re.sub(pattern, replace_var, result)
    
    return result


def _merge_configs(base: Dict[str, Any], override: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Recursively merge override into base config."""
    if not override:
        return base
    
    merged = dict(base)
    for key in override:
        if isinstance(merged.get(key), dict) and isinstance(override[key], dict):
            merged[key] = _merge_configs(merged[key], override[key])
        else:
            merged[key] = override[key]
    
    return merged


def parse_config_file(
    path: Union[str, Path],
    env_override: Optional[Dict[str, Any]] = None
) -> Config:
    """
    Parse a configuration file and return a validated Config object.
    
    Args:
        path: Path to the config file (YAML or JSON).
        env_override: Optional dictionary of environment variable overrides.
    
    Returns:
        A fully parsed and validated Config object.
    
    Raises:
        FileNotFoundError: If the config file does not exist.
        ValueError: If the config format is invalid.
    """
    path = Path(path)
    
    if not path.exists():
        raise FileNotFoundError(
            f"Config file not found: {path}"
        )
    
    # Load raw data
    try:
        with open(path) as f:
            raw_data = json.load(f)
    except json.JSONDecodeError:
        raw_data = _load_yaml_file(path)
    
    if not isinstance(raw_data, dict):
        raise ValueError(
            f"Config file must contain a JSON or YAML object, got {type(raw_data).__name__}"
        )
    
    # Expand environment variables in all string values
    def expand_strings(obj: Any) -> Any:
        if isinstance(obj, str):
            return _expand_env_vars(obj)
        elif isinstance(obj, dict):
            return {k: expand_strings(v) for k, v in obj.items()}
        elif isinstance(obj, list):
            return [expand_strings(item) for item in obj]
        else:
            return obj
    
    expanded_data = expand_strings(raw_data)
    
    # Merge environment overrides if provided
    merged_data = _merge_configs(expanded_data, env_override)
    
    # Parse into Config object
    try:
        config = Config(**merged_data)
    except TypeError as e:
        raise ValueError(f"Invalid config format: {e}") from e
    
    return config


def validate_config(config: Config) -> List[str]:
    """
    Perform additional validation checks on a parsed config.
    
    Returns a list of warning messages (empty if all checks pass).
    """
    warnings = []
    
    # Check for common issues
    if not config.targets.base_url:
        warnings.append("Warning: No base URL specified in targets")
    
    if not config.auth.type:
        warnings.append("Warning: Authentication type not specified; will use basic auth")
    
    if config.rate_limit.requests_per_second > 100.0:
        warnings.append(
            f"Warning: High request rate ({config.rate_limit.requests_per_second}/s) "
            "may cause target server issues"
        )
    
    # Check output directory exists or is writable
    try:
        Path(config.output.output_dir).mkdir(parents=True, exist_ok=True)
    except OSError as e:
        warnings.append(f"Warning: Output directory may not be writable: {e}")
    
    return warnings


def load_and_validate_config(
    path: Union[str, Path],
    env_override: Optional[Dict[str, Any]] = None
) -> Config:
    """
    Convenience function to load, parse, and validate a config file.
    
    Returns the parsed Config along with any warning messages.
    """
    try:
        config = parse_config_file(path, env_override)
        warnings = validate_config(config)
        
        if warnings:
            for w in warnings:
                print(f"  {w}")
        
        return config
    
    except (FileNotFoundError, ValueError) as e:
        print(f"Error loading config: {e}")
        raise


def create_default_config() -> Config:
    """Create a default configuration with sensible defaults."""
    warnings = validate_config(Config())
    
    if warnings:
        for w in warnings:
            print(f"  Default warning: {w}")
    
    return Config()


# ============================================================================
# Demo / Entry Point
# ============================================================================

if __name__ == "__main__":
    import sys
    
    # Example 1: Load from file (default demo config)
    print("=" * 60)
    print("Example 1: Loading default configuration")
    print("=" * 60)
    
    # Create a temporary config for demonstration
    temp_config = Path("/tmp/dastlite-demo-config.yaml")
    
    # Write a sample config file
    sample_yaml = """
# DASTLite Demo Configuration
targets:
  base_url: "https://example.com"
  paths:
    - "/"
    - "/api/v1"

auth:
  type: "bearer"
  token: "${DASTLITE_TOKEN:-demo-token-12345}"

ruleset:
  enabled: true
  categories:
    - xss
    - csrf
  severity_threshold: "high"

output:
  formats: ["sarif", "json"]
  output_dir: "./results"

rate_limit:
  requests_per_second: 25.0
  burst_size: 100
"""
    
    with open(temp_config, 'w') as f:
        f.write(sample_yaml)
    
    # Parse and display the config
    print(f"\nLoading from: {temp_config}")
    config = parse_config_file(temp_config)
    
    print("\nParsed Configuration:")
    print("-" * 40)
    print(json.dumps(config.to_dict(), indent=2))
    
    # Validate and show warnings
    print("\nValidation Warnings:")
    print("-" * 40)
    warnings = validate_config(config)
    if not warnings:
        print("  None (all checks passed)")
    else:
        for w in warnings:
            print(f"  - {w}")
    
    # Example 2: Load from command line argument or default
    print("\n" + "=" * 60)
    print("Example 2: Loading with environment override")
    print("=" * 60)
    
    env_override = {
        "targets": {"base_url": "https://api.example.com"},
        "auth": {"type": "basic", "username": "admin"}
    }
    
    config_overridden = parse_config_file(temp_config, env_override=env_override)
    print(f"\nOverridden base_url: {config_overridden.targets.base_url}")
    print(f"Overridden auth type: {config_overridden.auth.type}")
    
    # Example 3: Create default config
    print("\n" + "=" * 60)
    print("Example 3: Creating default configuration")
    print("=" * 60)
    
    default_config = create_default_config()
    print(f"\nDefault output directory: {default_config.output.output_dir}")
    print(f"Default ruleset enabled: {default_config.ruleset.enabled}")
    
    # Cleanup temp file
    if temp_config.exists():
        temp_config.unlink()
    
    print("\nDemo complete!")